{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***converting multi fast5 in mapped fastq to single fast5***\n",
    "\n",
    "Tasks we need to get: *Read ID of all mapped fastqs *extract them into batches *convert them into single fast5s\n",
    "\n",
    "**To perform the 1st task your input = mapped bam file multifast5 of fastqs that mapped to the reference genome make sure to define the input directories before running the codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "BAMIN_DIR = '../../analyses/mapping/infected_leaves/infected_leaves_3/'\n",
    "FAST5IN_DIR = '../../data/genomic_data/infected_leaves/workspace_fast5/infected_leaves_3_fast5_out/'\n",
    "OUT_DIR = '../../analyses/single_fast5s/infected_leaves/infectedleaves_3/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***Define directories by giving them absolute path instead of typing the pathway over and over again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "BAMIN_DIR = os.path.abspath(BAMIN_DIR)\n",
    "FAST5IN_DIR =os.path.abspath(FAST5IN_DIR)\n",
    "OUT_DIR =os.path.abspath(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "###catch the bam based on naming convention\n",
    "inbam_fn = [os.path.join(BAMIN_DIR, x) for x in os.listdir(BAMIN_DIR) if x.endswith ('sorted.bam')][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mappedids_fn = inbam_fn.replace('.bam', '.mappedids.txt')\n",
    "!samtools  view -F 4  {inbam_fn} | cut -f 1 | sort | uniq > {mappedids_fn}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***inorder to hold the batches of fast5s files generated by using the upper command, we need a temperory folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "TMPOUT_DIR = os.path.join(OUT_DIR, 'tmp')\n",
    "if not os.path.exists(TMPOUT_DIR):\n",
    "    os.makedirs(TMPOUT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***The files will be generated in a temperory folder, then to extract fast5s in batches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| 228603 of 228603|##########################################|100% Time: 2:19:03\n",
      "INFO:Fast5Filter:226661 reads extracted\n"
     ]
    }
   ],
   "source": [
    "###extract fast5s in bacthes\n",
    "subset_ids = os.path.basename(mappedids_fn).replace('.txt', '')\n",
    "!fast5_subset -i {FAST5IN_DIR} -s {TMPOUT_DIR} -f {subset_ids} -l {mappedids_fn} --recursive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filename_mapping.txt\r\n",
      "infected_leaves_3.sorted.mappedids0.fast5\r\n",
      "infected_leaves_3.sorted.mappedids10.fast5\r\n",
      "infected_leaves_3.sorted.mappedids11.fast5\r\n",
      "infected_leaves_3.sorted.mappedids12.fast5\r\n",
      "infected_leaves_3.sorted.mappedids13.fast5\r\n",
      "infected_leaves_3.sorted.mappedids14.fast5\r\n",
      "infected_leaves_3.sorted.mappedids15.fast5\r\n",
      "infected_leaves_3.sorted.mappedids16.fast5\r\n",
      "infected_leaves_3.sorted.mappedids17.fast5\r\n",
      "infected_leaves_3.sorted.mappedids18.fast5\r\n",
      "infected_leaves_3.sorted.mappedids19.fast5\r\n",
      "infected_leaves_3.sorted.mappedids1.fast5\r\n",
      "infected_leaves_3.sorted.mappedids20.fast5\r\n",
      "infected_leaves_3.sorted.mappedids21.fast5\r\n",
      "infected_leaves_3.sorted.mappedids22.fast5\r\n",
      "infected_leaves_3.sorted.mappedids23.fast5\r\n",
      "infected_leaves_3.sorted.mappedids24.fast5\r\n",
      "infected_leaves_3.sorted.mappedids25.fast5\r\n",
      "infected_leaves_3.sorted.mappedids26.fast5\r\n",
      "infected_leaves_3.sorted.mappedids27.fast5\r\n",
      "infected_leaves_3.sorted.mappedids28.fast5\r\n",
      "infected_leaves_3.sorted.mappedids29.fast5\r\n",
      "infected_leaves_3.sorted.mappedids2.fast5\r\n",
      "infected_leaves_3.sorted.mappedids30.fast5\r\n",
      "infected_leaves_3.sorted.mappedids31.fast5\r\n",
      "infected_leaves_3.sorted.mappedids32.fast5\r\n",
      "infected_leaves_3.sorted.mappedids33.fast5\r\n",
      "infected_leaves_3.sorted.mappedids34.fast5\r\n",
      "infected_leaves_3.sorted.mappedids35.fast5\r\n",
      "infected_leaves_3.sorted.mappedids36.fast5\r\n",
      "infected_leaves_3.sorted.mappedids37.fast5\r\n",
      "infected_leaves_3.sorted.mappedids38.fast5\r\n",
      "infected_leaves_3.sorted.mappedids39.fast5\r\n",
      "infected_leaves_3.sorted.mappedids3.fast5\r\n",
      "infected_leaves_3.sorted.mappedids40.fast5\r\n",
      "infected_leaves_3.sorted.mappedids41.fast5\r\n",
      "infected_leaves_3.sorted.mappedids42.fast5\r\n",
      "infected_leaves_3.sorted.mappedids43.fast5\r\n",
      "infected_leaves_3.sorted.mappedids44.fast5\r\n",
      "infected_leaves_3.sorted.mappedids45.fast5\r\n",
      "infected_leaves_3.sorted.mappedids46.fast5\r\n",
      "infected_leaves_3.sorted.mappedids47.fast5\r\n",
      "infected_leaves_3.sorted.mappedids48.fast5\r\n",
      "infected_leaves_3.sorted.mappedids49.fast5\r\n",
      "infected_leaves_3.sorted.mappedids4.fast5\r\n",
      "infected_leaves_3.sorted.mappedids50.fast5\r\n",
      "infected_leaves_3.sorted.mappedids51.fast5\r\n",
      "infected_leaves_3.sorted.mappedids52.fast5\r\n",
      "infected_leaves_3.sorted.mappedids53.fast5\r\n",
      "infected_leaves_3.sorted.mappedids54.fast5\r\n",
      "infected_leaves_3.sorted.mappedids55.fast5\r\n",
      "infected_leaves_3.sorted.mappedids56.fast5\r\n",
      "infected_leaves_3.sorted.mappedids5.fast5\r\n",
      "infected_leaves_3.sorted.mappedids6.fast5\r\n",
      "infected_leaves_3.sorted.mappedids7.fast5\r\n",
      "infected_leaves_3.sorted.mappedids8.fast5\r\n",
      "infected_leaves_3.sorted.mappedids9.fast5\r\n"
     ]
    }
   ],
   "source": [
    "!ls {TMPOUT_DIR}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***onces the reads are extracted, generated single fast5s from multifast5s of mapped reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| 57 of 57|##################################################|100% Time: 0:02:07\n"
     ]
    }
   ],
   "source": [
    "!multi_to_single_fast5 -i {TMPOUT_DIR} -s {OUT_DIR} -t 10 --recursive\n",
    "!touch {OUT_DIR}/single_fast5.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
